File: src/cmd/stop.go
package cmd

import (
	"fmt"

	"angel/src/core"
	"angel/src/core/config"
	"angel/src/core/launchctl"

	"github.com/alecthomas/kong"
)

type StopCmd struct {
	Name string `arg:"" help:"Service name to stop."`
}

func (s *StopCmd) Run(a *core.Angel, config *config.Config, ctx *kong.Context) error {
	return a.Daemons.WithMatch(s.Name, false, ctx, func(daemon core.Daemon) error {
		output, err := launchctl.Kill(daemon)
		if err != nil {
			return err
		}
		fmt.Print(string(output))
		fmt.Printf("stopped %s\n", daemon.Name)
		return nil
	})
}

File: src/commands/stop.rs
use crate::angel::Angel;
use crate::cli::StopArgs;
use crate::error::Result;
use crate::launchctl;
use crate::output::stdout;

pub fn run(angel: &Angel, args: &StopArgs) -> Result<()> {
    let daemon = angel.daemons.get_match(&args.name, args.exact)?;
    let result = launchctl::kill(daemon, &args.signal.to_string())?;
    stdout::write(&result.output);
    match result.success() {
        true => stdout::success(&format!("stopped {}", daemon.name)),
        false => stdout::error(&format!(
            "failed to stop {}: {}",
            daemon.name, result.stderr
        )),
    }
    Ok(())
}
